### ISSUE TITLE
Problem with directory upload

### ISSUE THREAD

**Issue by RenatoPerotti** [ipfs-shipyard/ipfs-webui#872]
The '+Add to IPFS' button in 'Files' section, does not work for uploading a directory.

**Comment by lidel** [comment]
@RenatoPerotti are you able to provide more context?

1. What are contents of the directory? Is there an error?
1. What is the version of Web UI you use? (in left bottom corner)
1. What is the node type you use? (direct go-ipfs? js-ipfs? proxied via ipfs-companion?)
1. Which option do you use? 
   > ![2018-10-28--17-45-35](https://user-images.githubusercontent.com/157609/47618971-6742f580-dad9-11e8-8792-bb5e2201c5d5.png)

**Comment by hacdias** [comment]
@RenatoPerotti also, which browser are you using? (The directory upload functionality on browsers is a bit limited and we've only tested Firefox and Chrome).

**Comment by RenatoPerotti** [comment]
Hi,
I tried it on 2 machines... one windows using Firefox 63.0 64bit (companion installed)... the other is a quite fresh Ubuntu setup with its standard Firefox (without companion).
Both the nodes are jsipfs daemon's, and the error shows "Failed to add files to IPFS. Please try again." on the bottom in orange. I used the "add folder" button, and it showed "100%" for a short period.
The content that I tried is a folder with just 1 jpg image (and the jpg on its own, uploads fine)

I hope this info helps, but feel free to get in contact if you need more answers :)

**Comment by lidel** [comment]
Thank you for reporting this @RenatoPerotti !


I was able to reproduce the issue with:
-  ipfs-webui v2.1.1
-  both js-ipfs (v0.32.3) and go-ipfs (0.4.17 and 0.4.18-rc1)

Turns out the data is added to IPFS correctly, but Web UI fails to recognize success and wisplays error instead of adding CID to MFS:

> #### successful `/api/v0/add` with go-ipfs:
> ![noop-goipfs-0 4 18-2018-10-31--12-32-40](https://user-images.githubusercontent.com/157609/47785702-c7b87980-dd09-11e8-94e8-fd14b837a46a.png)
> 
> #### successful `/api/v0/add`  with js-ipfs:
> ![noo-upload-2018-10-31--12-23-52](https://user-images.githubusercontent.com/157609/47785696-c2f3c580-dd09-11e8-963e-ab0a73b53ce6.png)
> 
> #### false-negative result in ipfs-webui:
> 
> ![2018-10-31--12-33-20](https://user-images.githubusercontent.com/157609/47785662-abb4d800-dd09-11e8-8bd7-2a024677ab3a.png)
> 

@hacdias  turns out this false-negative is a side-effect of this fix: 

https://github.com/ipfs-shipyard/ipfs-webui/blob/bade669b91c63b050612028fc6528386a44b6019/src/bundles/files.js#L244-L246 

I think we need to make it smarter so it does expect additional wrapping directory when we "Add directory" because it is already there.

**Comment by RenatoPerotti** [comment]
I tried to look for the part in this web-UI that uploads a merkle-tree (for files) and found a piece of code that should do something I guess. I have removed the async parts to try it out in the browser console. 

function makeHashFromFiles (files, ipfs) {
  let node = ipfs.object.new('unixfs-dir')

  for (const file of files) {
    node = ipfs.object.patch.addLink(node.toJSON().multihash, {
      name: file.name,
      size: file.size,
      multihash: file.hash
    })
  }
}

It does need a declaration of the "files" property I would think... Can anyone show how to do this?

Thanx
Let's boldly go... IPFS :)
